globus_automate_client/cli/rich_helpers.py

```diff
@@ -0,0 +1,435 @@
+    prehook = functools.partial(identity_to_user, "created_by")
+
+
+class RunEnumerateDisplayFields(RunListDisplayFields):
+    """
+    This object defines the fields and display style of a Run enumeration into a
+    table.
+    """
+
+    path_to_data_list = "runs"
+
+
+class FlowListDisplayFields(DisplayFields):
+    """
+    This object defines the fields and display style of a Flow listing into a
+    table.
+    """
+
+    fields = [
+        Field("title", ""),
+        Field("id", ""),
+        Field("flow_owner", "", humanize_auth_urn),
+        Field("created_at", "", humanize_datetimes),
+        Field("updated_at", "", humanize_datetimes),
+    ]
+    path_to_data_list = "flows"
+    prehook = functools.partial(identity_to_user, "flow_owner")
+
+
+class RunLogDisplayFields(DisplayFields):
+    """
+    This object defines the fields and display style of a Run log listing into a
+    table.
+    """
+
+    fields = [
+        Field("code", ""),
+        Field("description", ""),
+        Field("time", "", humanize_datetimes),
+    ]
+    path_to_data_list = "entries"
+
+
+class CompletionDetetector(abc.ABC):
+    """
+    An object that can be used to determine if an operation is complete or if it
+    should continue polling.
+    """
+
+    terminals: Set[str]
+
+    @classmethod
+    @abc.abstractmethod
+    def is_complete(self, result: Union[GlobusHTTPResponse, GlobusAPIError]) -> bool:
+        """
+        Given either a GloubsHTTPReponse or GlobusAPIError, this method should
+        return a boolean indicating if polling should continue.
+        """
+        pass
+
+
+class ActionCompletionDetector(CompletionDetetector):
+    """
+    This class determines when a Run has reached a completed state.
+    """
+
+    terminals: Set[str] = {"SUCCEEDED", "FAILED"}
+
+    @classmethod
+    def is_complete(self, result: Union[GlobusHTTPResponse, GlobusAPIError]) -> bool:
+        return (
+            isinstance(result, GlobusAPIError)
+            or result.data.get("status", None) in self.terminals
+        )
+
+
+class LogCompletionDetetector(CompletionDetetector):
+    """
+    This class determines when a Run has reached a completed state from
+    inspecting its logs.
+    """
+
+    terminals: Set[str] = {"FlowSucceeded", "FlowFailed", "FlowCanceled"}
+
+    @classmethod
+    def is_complete(self, result: Union[GlobusHTTPResponse, GlobusAPIError]) -> bool:
+        return (
+            isinstance(result, GlobusAPIError)
+            or result.data["entries"][-1]["code"] in self.terminals
+        )
+
+
+class Result:
+    """
+    A class which wraps a Globus API response.
+    """
+
+    def __init__(
+        self,
+        response: Union[GlobusHTTPResponse, GlobusAPIError, str],
+        detetector: Optional[Type[CompletionDetetector]] = None,
+    ):
+        self.result = response
+        if detetector is None:
+            self.detetector: Type[CompletionDetetector] = ActionCompletionDetector
```

Comment:
Could this just be the default on the method rather than making it `Optional` and then testing for `None`?

Reply:
Good suggestion for simplifying the interface. Made the change + propagated it out.